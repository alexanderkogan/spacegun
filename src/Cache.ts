export class Cache<S, T> {
    private data: Map<S, { value: T, age: number }> = new Map()

    public constructor(
        public maxAgeSeconds: number,
        public maxSize: number = 100
    ) { }

    public get(key: S): T | undefined {
        const entry = this.data.get(key)
        if (entry !== undefined && Date.now() - entry.age < this.maxAgeSeconds * 1000) {
            return entry.value
        }
        return undefined
    }

    public set(key: S, value: T): void {
        this.data.set(key, { value, age: Date.now() })
    }

    public async calculate(key: S, valueProvider: () => Promise<T>): Promise<T> {
        let value = this.get(key)
        if (value === undefined) {
            value = await valueProvider()
            this.set(key, value)
        }
        return value
    }
}
