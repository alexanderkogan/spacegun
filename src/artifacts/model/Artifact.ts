export interface Artifact {
    name: string
    data: object
}
