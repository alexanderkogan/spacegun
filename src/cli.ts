import { parse } from "./options"
import { KubernetesClusterProvider } from './cluster/kubernetes/KubernetesCluster'
import { commands } from "./commands"

(async () => {
    const options = parse()
    const cluster = new KubernetesClusterProvider(options.kube)
    try {
        await commands[options.command](cluster)
    } catch (error) {
        console.error(`Command ${options.command} failed: ${error}`)
    }
})()
