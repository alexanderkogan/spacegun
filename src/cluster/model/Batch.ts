import { Image } from './Image'

export interface Batch {
    name: string
    image?: Image
}
