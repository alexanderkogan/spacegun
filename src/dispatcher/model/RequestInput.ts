import { Params } from "./Params"

export class RequestInput {
    constructor(
        public params?: Params,
        public data?: any
    ) { }

    public static ofData(data: any, ...p: [string, any][]): RequestInput {
        return new RequestInput(RequestInput.buildParams(p), data)
    }

    public static of(...p: [string, any][]): RequestInput {
        return new RequestInput(RequestInput.buildParams(p))
    }

    private static buildParams(p: [string, any][]): Params | undefined {
        if (p.length > 0) {
            const params: Params = {}
            p.filter(([{ }, value]) => value !== undefined).forEach(([key, value]) => {
                const current = params[key]
                const v = value.toString()
                if (current === undefined) {
                    params[key] = v
                } else if (typeof current === "string") {
                    params[key] = [current, v]
                } else {
                    current.push(v)
                }
            })
            return params
        } else {
            return undefined
        }
    }
}
