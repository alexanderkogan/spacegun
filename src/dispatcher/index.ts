
import { PromiseProvider } from "@/dispatcher/model/PromiseProvider"
import { RequestInput } from "@/dispatcher/model/RequestInput"
import { build, listen, init } from "@/dispatcher/api"

let procedures: {
    [name: string]: PromiseProvider<any, any>
} = {}

export function add<T>(moduleName: string, procedureName: string, procedure: PromiseProvider<RequestInput, T>) {
    const p = path(moduleName, procedureName)
    if (procedures[p] === undefined) {
        procedures[p] = procedure
    } else {
        throw Error(`duplicate registration of procedure ${p}`)
    }
}

export function get<T>(moduleName: string, procedureName: string): PromiseProvider<RequestInput, T> {
    return procedures[path(moduleName, procedureName)]
}

export function reset(): void {
    procedures = {}
}

export function path(moduleName: string, procedureName: string): string {
    return `${moduleName}/${procedureName}`
}

export function run(serverHost: string, serverPort: number): void {
    init(serverHost, serverPort)
    build()
    listen()
}
