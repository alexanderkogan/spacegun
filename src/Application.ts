import { load as loadConfig, Config } from "@/config"
import { commands, printHelp } from "@/commands"
import { parse } from "@/options"

import { Layers } from "@/dispatcher/model/Layers"
import { run as runDispatcher } from "@/dispatcher"

import { IO } from "@/IO"
import { CronRegistry } from "@/crons/CronRegistry"

import { GitConfigRepository, fromConfig as gitRepoFromConfig } from "@/config/git/GitConfigRepository"
import { fromConfig as configRepoFromConfig } from "@/config/filesystem/FilesystemConfigRepository";

import { KubernetesClusterRepository } from "@/cluster/kubernetes/KubernetesClusterRepository"
import { DockerImageRepository } from "@/images/docker/DockerImageRepository"
import { JobsRepositoryImpl } from "@/jobs/JobsRepositoryImpl"
import { SlackEventRepository } from "@/events/slack/SlackEventRepository"

import { init as initCluster } from "@/cluster/ClusterModule"
import { init as initEvents } from "@/events/EventModule"
import { init as initImages } from "@/images/ImageModule"
import { init as initConfig } from "@/config/ConfigModule"
import { init as initJobs } from "@/jobs/JobsModule"
import { init as initViews } from "@/views"

import { Options } from "@/options"

export class Application {

    public static create(): Application {
        const io = new IO()
        const crons = new CronRegistry()
        const options = parse()
        return new Application(io, crons, options)
    }

    public constructor(
        public readonly io: IO,
        public readonly crons: CronRegistry,
        public readonly options: Options
    ) { }

    public async run() {
        try {
            const config = loadConfig(this.options.config)
            this.initialize(config)

            runDispatcher(config.server.host, config.server.port)
            if (process.env.LAYER === Layers.Standalone || process.env.LAYER === Layers.Client) {
                await commands[this.options.command](this.io)
            }
        } catch (e) {
            printHelp(this.io, e)
        }
    }

    public async checkForConfigChange(git: GitConfigRepository) {
        const hasNewConfig = await git.hasNewConfig()
        if (hasNewConfig) {
            await git.fetchNewConfig()
            this.crons.removeAllCrons()
            this.reload()
        }
    }

    private reload(): void {
        try {
            const config = loadConfig(this.options.config)
            this.initialize(config)
        } catch (e) {
            this.io.out(`could not reload config ${e.message}`)
        }
    }

    private initialize(config: Config) {
        const gitRepo = gitRepoFromConfig(config)
        if (gitRepo !== undefined) {
            this.crons.register(
                "config-reload",
                config.git!.cron,
                () => this.checkForConfigChange(gitRepo)
            )
            initConfig(gitRepo)
        } else {
            const fileSystemRepo = configRepoFromConfig(config)
            initConfig(fileSystemRepo)
        }

        initViews(config)
        initEvents([
            SlackEventRepository.fromConfig(config.slack)
        ])
        initCluster(KubernetesClusterRepository.fromConfig(config.kube, config.namespaces))
        initImages(DockerImageRepository.fromConfig(config.docker))
        initJobs(JobsRepositoryImpl.fromConfig(config.jobs, this.crons))
    }
}
