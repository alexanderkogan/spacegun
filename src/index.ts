import { Layers } from "@/dispatcher/model/Layers"

import { load as loadConfig } from "@/config"
import { parse } from "@/options"
import { run as runDispatcher } from "@/dispatcher"
import { commands, printHelp } from "@/commands"
import { IO } from "@/IO"

import { init as initJobs } from "@/jobs/JobsModule"
import { init as initCluster } from "@/cluster/ClusterModule"
import { init as initImages } from "@/images/ImageModule"

import { KubernetesClusterRepository } from "@/cluster/kubernetes/KubernetesClusterRepository"
import { DockerImageRepository } from "@/images/docker/DockerImageRepository"
import { JobsRepositoryImpl } from "@/jobs/JobsRepositoryImpl"

const options = parse()
const config = loadConfig(options.config)
const io = new IO()
try {
    if (config instanceof Error) {
        printHelp(io, true)
    } else {
        initJobs(JobsRepositoryImpl.fromConfig(config.jobs))
        initCluster(KubernetesClusterRepository.fromConfig(config.kube))
        initImages(DockerImageRepository.fromConfig(config.docker))

        runDispatcher(config.server.host, config.server.port)
        if (process.env.LAYER === Layers.Standalone || process.env.LAYER === Layers.Client) {
            commands[options.command](io)
        }
    }
} catch (e) {
    console.error(e)
}
