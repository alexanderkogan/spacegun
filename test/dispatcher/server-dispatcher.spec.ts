import { callParameters } from "../test-utils/jest"

import { get } from "../../src/dispatcher"

process.env.LAYER = "server"
import { moduleName, globalFunction, remoteFunction, functions } from "./TestModule"
import { RequestInput } from "../../src/dispatcher/model/RequestInput";

describe("server dispatcher", () => {

    beforeEach(() => {
        jest.resetAllMocks()
    })

    it("calls global functions locally", () => {
        // when
        get(moduleName, functions.standalone)()

        // then
        expect(globalFunction).toHaveBeenCalledTimes(1)
    })

    it("throws error on calls to local functions", () => {
        expect(() => get(moduleName, functions.local)()).toThrowErrorMatchingSnapshot()
    })

    it("throws error on calls to local functions with parameters", () => {
        expect(() => get(moduleName, functions.local)(RequestInput.of(["param", 1]))).toThrowErrorMatchingSnapshot()
    })

    it("calls remote void function locally", async () => {
        // when
        await get(moduleName, functions.remoteVoid)()

        // then
        expect(remoteFunction).toHaveBeenCalled()
    })

    it("calls remote params function locally", async () => {
        // given
        const params = RequestInput.of(["param", 1])

        // when
        await get(moduleName, functions.remoteParams)(params)

        // then
        expect(remoteFunction).toHaveBeenCalled()
    })

    it("extracts params from remote params call", async () => {
        // given
        const params = RequestInput.of(["param", 1], ["param", 2], ["another param", "1"])

        // when
        await get(moduleName, functions.remoteParams)(params)

        // then
        expect(callParameters(remoteFunction)[0]).toEqual(1)
    })
})
