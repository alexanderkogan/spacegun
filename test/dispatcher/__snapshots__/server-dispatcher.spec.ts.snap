// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server dispatcher throws error on calls to local functions 1`] = `"a call to clientModule/local with input {} in layer client from server is not possible"`;

exports[`server dispatcher throws error on calls to local functions with parameters 1`] = `"a call to clientModule/local with input {\\"params\\":{\\"param\\":\\"1\\"}} in layer client from server is not possible"`;
